Rewrite frame CRC generation

Starkku opened the pull request:
Reimplements game's frame CRC generation function, only change logic-wise is that it now excludes `AnimClass` and `ParticleClass` objects that are purely visual.

Starkku committed:
Rewrite frame CRC generation

Starkku committed:
Rewrite frame CRC generation

FS-21 commented:
What kind of tests should be done here if you want feedback?

chaserli commented on src/Misc/SyncLogging.cpp:
```diff
@@ -527,3 +530,96 @@ DEFINE_HOOK(0x683AB0, ScenarioClass_Start_DisableSyncLog, 0x6)
+		if (pObj->UniqueID == -2)
+			return true;
+
+		auto const pAnim = abstract_cast<AnimClass*>(pObj);
```
```suggestion
		auto const pAnim = static_cast<AnimClass*>(pObj);
```

chaserli commented on src/Misc/SyncLogging.cpp:
```diff
@@ -527,3 +530,96 @@ DEFINE_HOOK(0x683AB0, ScenarioClass_Start_DisableSyncLog, 0x6)
+	}
+	else if (rtti == AbstractType::Particle)
+	{
+		auto const pParticle = abstract_cast<ParticleClass*>(pObj);
```
```suggestion
		auto const pParticle = static_cast<ParticleClass*>(pObj);
```

chaserli commented on src/Misc/SyncLogging.cpp:
```diff
@@ -527,3 +530,96 @@ DEFINE_HOOK(0x683AB0, ScenarioClass_Start_DisableSyncLog, 0x6)
+{
+	enum { SkipGameCode = 0x64DE7C };
+
+	auto& GameCRC = EventClass::CurrentFrameCRC.get();
```
```suggestion
	auto& GameCRC = EventClass::CurrentFrameCRC();
```

chaserli reviewed (changes requested):
If you're rewriting these completely, it would be better to hash these in a more modern way for better readability and extensibility

ZivDero commented on src/Misc/SyncLogging.cpp:
```diff
@@ -527,3 +530,141 @@ DEFINE_HOOK(0x683AB0, ScenarioClass_Start_DisableSyncLog, 0x6)
+
+#pragma region mp_frame_crc
+
+static bool __forceinline ObjectNonCRCHashable(ObjectClass const* pObj)
```
```cpp
class ObjectClassFake : public ObjectClass
{
	inline bool _IsCRCHashable();
}

bool ObjectClassFake::_IsCRCHashable
{
...
}
```

ZivDero replied:
Instead of creating global functions where one would normally create a class method, you could create a "fake" class descending from ObjectClass, allowing you to effectively add new (non-virtual, of course) class methods, with access to `this`.

ZivDero commented on src/Misc/SyncLogging.cpp:
```diff
@@ -527,3 +530,141 @@ DEFINE_HOOK(0x683AB0, ScenarioClass_Start_DisableSyncLog, 0x6)
+	}
+};
+
+DEFINE_HOOK(0x64DAB0, ComputeFrameCRC, 0x6)
```
Please reimplement the whole function as a normal (non-hook) function and patch it with a simple jump for cleanliness.

ZivDero commented on src/Misc/SyncLogging.cpp:
```diff
@@ -527,3 +530,141 @@ DEFINE_HOOK(0x683AB0, ScenarioClass_Start_DisableSyncLog, 0x6)
+
+#define FRAMECRC_HASH_INI
+
+#ifdef FRAMECRC_HASH_INI
```
No.

Metadorius replied:
No.

Starkku committed:
Improvements

- Use DEFINE_JUMP to replace calls to ComputeGameCRC instead of hooking in it
- Separate the CRC calculations into inlined functions
- Do not inline the hash eligibility check function due to its size, rename it for clarity
- Misc. minor improvements

Starkku committed:
Improvements

- Use DEFINE_JUMP to replace calls to ComputeGameCRC instead of hooking in it
- Separate the CRC calculations into inlined functions
- Do not inline the hash eligibility check function due to its size, rename it for clarity
- Misc. minor improvements

Starkku commented:
The wrapper may be good idea in theory but the current implementation mostly just obfuscates what gets hashed and for what type of object. The benefits would be more obvious if there was a realistic chance for new properties that need hashing to be added for, say, TechnoClass and derivatives in the future but as it stands this is *very* unlikely. If you can think of a way to do it without extra overhead or the obfuscating nature while reducing repeated code be my guest, but the previous implementation was not it so I've elected to revert it. I do hold a local copy of it for now still just in case though.

I'm not even going to repeat the point about the rules hash check in Phobos anymore, should be obvious by now that it does not belong here.

I did do some minor cleanup to the first iteration.

secsome commented on docs/Whats-New.md:
```diff
@@ -451,6 +451,7 @@ New:
 - Nonprovocative Warheads (by Starkku)
 - Option to restore `PowerSurplus` setting for AI (by Starkku)
 - `FireOnce` infantry sequence reset toggle (by Starkku)
+- Purely visual animations and particles excluded from sync checks (by Starkku)
```
I feel like this PR is more like a `Vanilla Fix` instead of a `New Feature`

chaserli commented:
> If you can think of a way to do it without extra overhead or the obfuscating nature while reducing repeated code be my guest
It only wrapped a DWORD and serves for hash the object on creation and do the hash combine operation. I bet it makes less overhead than your current method. It also allows you to know the "sum" at compile time. But anyway these are small things
 and make literally no difference

Starkku committed:
Fix errors

Starkku committed:
Fix errors

Starkku committed:
Improvements & fixes

Starkku committed:
Improvements & fixes

Starkku commented:
Rebased on develop and applied some of the feedback. This is ready for further testing and another review pass.

Coronia commented:
no desync or other issue occured in our testing so far

Starkku committed:
Consider Next(Particle) when checking eligibility of animation/particle for hashing

Starkku committed:
Consider Next(Particle) when checking eligibility of animation/particle for hashing

Starkku committed:
Fix issue with NextParticle check leading to crashes

Starkku committed:
Fix issue with NextParticle check leading to crashes

ZivDero commented on src/Misc/SyncLogging.cpp:
```diff
@@ -541,3 +542,128 @@ DEFINE_HOOK(0x7013A0, TechnoClass_OverrideMission_SyncLog, 0x5)
+		if (this->UniqueID == -2)
+			return false;
+
+		auto const pAnim = static_cast<AnimClass*>((ObjectClass*)this);
```
Uh.... use `reinterpret_cast`? Or two `static_cast`, not a C cast.

ZivDero commented on src/Misc/SyncLogging.cpp:
```diff
@@ -541,3 +542,128 @@ DEFINE_HOOK(0x7013A0, TechnoClass_OverrideMission_SyncLog, 0x5)
+{
+	return location.X / 10 + ((location.Y / 10) << 16);
+}
+
```
Any real need for `__forceinline`? Just mark it `inline`, should be more than enough. High chance it will get inlined even without either.

Metadorius replied:
doesn't `inline` have different semantcis nowadays?

ZivDero replied:
Like what? Not that I'm aware.

Metadorius replied:
> The original intent of the inline keyword was to serve as an indicator to the optimizer that inline substitution of a function is preferred over function call, that is, instead of executing the function call CPU instruction to transfer control to the function body, a copy of the function body is executed without generating the call. This avoids overhead created by the function call (passing the arguments and retrieving the result) but it may result in a larger executable as the code for the function has to be repeated multiple times.

> Since inline substitution is unobservable in the standard semantics, compilers are free to use inline substitution for any function that's not marked inline, and are free to generate function calls to any function marked inline. Those optimization choices do not change the rules regarding multiple definitions and shared statics listed above.

> Because the meaning of the keyword inline for functions came to mean "multiple definitions are permitted" rather than "inlining is preferred" since C++98, that meaning was extended to variables.

Starkku replied:
Realistically could probably remove both keywords and let the compiler do its thing. As if I know better than it.

ZivDero reviewed (approved):
A couple nitpicks but looks good, see no issues.

Starkku committed:
Address feedback, add comments for clarity

Starkku committed:
Do not use inline keyword, compiler can do compiler things
